pkg/drawio/layout.go

```diff
@@ -0,0 +1,298 @@
+// setNetworkIconsLocations() sets all the icons in the first col.
+// choose the rows with heights >= iconSpace, and the rows next to them
+func (ly *layoutS) setNetworkIconsLocations() {
+	icons := ly.network.IconTreeNodes()
+	if len(icons) == 0 {
+		return
+	}
+	rows := []*row{}
+	for ri, row := range ly.matrix.rows {
+		if row.height() >= iconSpace {
+			rows = append(rows, row)
+			if nextRow := ly.matrix.rows[ri+1]; nextRow.height() < iconSpace {
+				rows = append(rows, nextRow)
+			}
+		}
+	}
+	iconsPerRow := (len(icons)-1)/len(rows) + 1
+	ly.network.Location().firstCol.setWidth(iconSpace * iconsPerRow)
+	for iconIndex, icon := range icons {
+		icon.setLocation(newCellLocation(rows[iconIndex/iconsPerRow], ly.matrix.cols[0]))
+		icon.Location().xOffset = iconSpace*(iconIndex%iconsPerRow) - (iconSpace*(iconsPerRow-1))/2
+	}
+}
+
+// ////////////////////////////////////////////////////////////////////////////////////////
+// setVpcIconsLocations() sets all the icons in the first vpc row.
+// choose the cols with wight >= iconSpace, and the cols below them
+func (ly *layoutS) setVpcIconsLocations(vpc SquareTreeNodeInterface) {
+	icons := vpc.IconTreeNodes()
+	if len(icons) == 0 {
+		return
+	}
+
+	cols := []*col{}
+	firstColIndex := vpc.Location().firstCol.index
+	lastColIndex := vpc.Location().lastCol.index
+	for ci := firstColIndex; ci <= lastColIndex; ci++ {
+		col := ly.matrix.cols[ci]
+		if col.width() >= iconSpace {
+			cols = append(cols, col)
+			if nextCol := ly.matrix.cols[ci+1]; nextCol.width() < iconSpace {
+				cols = append(cols, nextCol)
+			}
+		}
+	}
+	iconsPerCol := (len(icons)-1)/len(cols) + 1
+	if vpc.Location().firstRow.height() < iconSpace*iconsPerCol {
+		vpc.Location().firstRow.setHeight(iconSpace * iconsPerCol)
+	}
+	for iconIndex, icon := range icons {
+		icon.setLocation(newCellLocation(vpc.Location().firstRow, cols[iconIndex/iconsPerCol]))
+		icon.Location().yOffset = iconSpace*(iconIndex%iconsPerCol) - (iconSpace*(iconsPerCol-1))/2
+	}
+}
+
+// ////////////////////////////////////////////////////////////////////////////////////////
+// if vsi icon shares by several subnet - we put it below one of the subnets
+// else we put it inside the subnet
+// gateway we put at the top
+
+func (ly *layoutS) setZoneIconsLocations(zone SquareTreeNodeInterface) {
+	icons := zone.IconTreeNodes()
+	for _, icon := range icons {
+		if icon.IsVSI() {
+			vsiIcon := icon.(*VsiTreeNode)
+			vsiSubnets := vsiIcon.GetVsiSubnets()
+			if len(vsiSubnets) == 1 {
+				icon.setParent(vsiIcon.nis[0].Parent())
+				nisCombinedLocation := mergeLocations(locations(vsiIcon.nis))
+				icon.setLocation(newCellLocation(nisCombinedLocation.firstRow, nisCombinedLocation.firstCol))
+				if nisCombinedLocation.firstRow == nisCombinedLocation.lastRow {
+					vsiIcon.Location().yOffset = iconSize
+				} else {
+					vsiIcon.Location().yOffset = subnetHeight / 2
+				}
```

Comment:
This code needs some commenting - it's hard to understand its intent

Reply:
done